### PR TITLE
parse gtid for parseHandler if MySQL works in GTID_MODE, and begin to startWithGTID after mysqldump is done

### PR DESCRIPTION
**Pull request by jianhaiqing**
Purpose of This PR:
1. if MySQL is GTID_MODE=ON, canal would be start to replicate with GTID after mysqldump is done.
2. Add one more func GtidSet() for parseHandler, to save gtid set after parsing gtid from mysqldump.

### CODE REVIEW

**Comment by siddontang** [comment]
PTAL @GregoryIan

**Comment by siddontang** [comment]
CI failed @jianhaiqing

**Comment by jianhaiqing** [comment]
More details about GTID between MySQL  and MariaDB.
mysqldump --set-gtid-purged from MySQL  is not compatible with mysqldump of MariaDB, while MariaDB use --gtid instead
```
 [MariaDB]
 mysqldump --host=`hostname -i` --port=3306 --user=root --password=my-secret-pw --master-data --gtid --single-transaction --skip-lock-tables --compact --skip-opt --quick --no-create-info --skip-extended-insert --skip-tz-utc --hex-blob --default-character-set=utf8 longrun t6

-- CHANGE MASTER TO MASTER_LOG_FILE='mariadb-bin.000001', MASTER_LOG_POS=1138;
CHANGE MASTER TO MASTER_USE_GTID=slave_pos;
SET GLOBAL gtid_slave_pos='0-1-4';
INSERT INTO `t6` VALUES (1,11);
INSERT INTO `t6` VALUES (2,11);

[MySQL]
mysqldump --host=`hostname -i` --port=23306 --user=root --password=my-secret-pw --master-data --set-gtid-purged --single-transaction --skip-lock-tables --compact --skip-opt --quick --no-create-info --skip-extended-insert --skip-tz-utc --hex-blob --default-character-set=utf8 longrun t6

SET @MYSQLDUMP_TEMP_LOG_BIN = @@SESSION.SQL_LOG_BIN;
SET @@SESSION.SQL_LOG_BIN= 0;
SET @@GLOBAL.GTID_PURGED='5906b63d-4009-11e9-81f3-00505690245b:1-144473,
abc8b882-b168-11e8-a790-1866daabb9a4:1-3439091213';
CHANGE MASTER TO MASTER_LOG_FILE='mysql-bin.000258', MASTER_LOG_POS=4372;
INSERT INTO `t6` VALUES (1,11);
INSERT INTO `t6` VALUES (2,11);
INSERT INTO `t6` VALUES (3,11);
SET @@SESSION.SQL_LOG_BIN = @MYSQLDUMP_TEMP_LOG_BIN;
```

File: canal/canal.go [hunk 1]
```diff
@@ -120,6 +120,23 @@ func NewCanal(cfg *Config) (*Canal, error) {
...
+func (c *Canal) validateSetGtidPurged() error {
+	gtidPuged := strings.ToLower(c.cfg.Dump.GtidPurged)
+	if gtidPuged == "none" {
+		return nil
+	} else if gtidPuged == "auto" {
+		res, err := c.Execute(`SHOW GLOBAL VARIABLES LIKE "gtid_mode";`)
+		if err != nil {
+			return errors.Trace(err)
+		} else if f, _ := res.GetString(0, 1); f != "ON" {
+			return errors.Errorf("set-gtid-purged: %s, gtid_mode should be ON %s, but now is ", c.cfg.Dump.GtidPurged, f)
```
Comment (siddontang) [inline canal/canal.go]:
```
gtid_mode should be ON %s, but now is 
```
wrong place for `%s` verb?

File: replication/binlogsyncer.go [hunk 2]
```diff
@@ -496,6 +498,12 @@ func (b *BinlogSyncer) writeBinlogDumpMariadbGTIDCommand(gset GTIDSet) error {
 func (b *BinlogSyncer) localHostname() string {
 	if len(b.cfg.Localhost) == 0 {
 		h, _ := os.Hostname()
+		// if os.Hostname() lengther than 60, use md5sum instead
+		if len(h) > 60 {
```
Comment (siddontang) [inline replication/binlogsyncer.go]: why md5? 

seem this change is not relevant to this PR.

File: dump/parser.go [hunk 3]
```diff
@@ -54,19 +65,57 @@ func Parse(r io.Reader, h ParseHandler, parseBinlogPos bool) error {
...
+			}
+		}
+
 		if parseBinlogPos && !binlogParsed {
 			if m := binlogExp.FindAllStringSubmatch(line, -1); len(m) == 1 {
 				name := m[0][1]
 				pos, err := strconv.ParseUint(m[0][2], 10, 64)
 				if err != nil {
-					return errors.Errorf("parse binlog %v err, invalid number", line)
+					errors.Errorf("parse binlog %v err, invalid number", line)
```
Comment (siddontang) [inline dump/parser.go]: meaningless for this line if no return

File: replication/binlogsyncer.go (hunk 2 above)
Comment (jianhaiqing) [reply replication/binlogsyncer.go]: Current situation is that my company's  go-mysql-elasticsearch runs in K8S, the hostname set by our deliver system is longer than 62. And it's not easy to add an cfg.localhost for go-mysql-elasticsearch. I can remove these code  in this PR.
Is there suggestion for this kind of situation like K8S ,hostname is very long.

File: canal/canal.go (hunk 1 above)
Comment (jianhaiqing) [reply canal/canal.go]: yes, I fix it now

File: dump/parser.go (hunk 3 above)
Comment (jianhaiqing) [reply dump/parser.go]: add return  for each errors.Errorf

File: canal/config.go [hunk 4]
```diff
@@ -34,6 +34,9 @@ type DumpConfig struct {
 	// 'FLUSH TABLES WITH READ LOCK'
 	SkipMasterData bool `toml:"skip_master_data"`
 
+	// set --set-gtid-purged none, auto; none for gtid is disabled or "version too low", auto for gtid_mode=on;
```
Comment (WangXiangUSTC) [inline canal/config.go]: why when set to `auto` then enable gtid_mode? how about use `on` or `off`, or use bool type

File: canal/canal.go [hunk 5]
```diff
@@ -120,6 +120,23 @@ func NewCanal(cfg *Config) (*Canal, error) {
...
+		res, err := c.Execute(`SHOW GLOBAL VARIABLES LIKE "gtid_mode";`)
+		if err != nil {
+			return errors.Trace(err)
+		} else if f, _ := res.GetString(0, 1); f != "ON" {
+			return errors.Errorf("set-gtid-purged: %s, gtid_mode should be ON, but now is %s", c.cfg.Dump.GtidPurged, f)
+		}
+		return nil
+	}
+
+	return errors.Errorf("set_gtid_purged: on OR auto  can be set, current is %s", gtidPuged)
```
[the remark this answers is not in the document]
Comment (jianhaiqing) [reply canal/canal.go]: 'auto' or 'none' can be set.

File: canal/canal.go [hunk 6]
```diff
@@ -128,6 +145,12 @@ func (c *Canal) prepareDumper() error {
 		return nil
 	}
 
+	// validate c.cfg.Dump.GtidPurged)
```
[the remark this answers is not in the document]
Comment (jianhaiqing) [reply canal/canal.go]: ok

File: canal/config.go (hunk 4 above)
Comment (jianhaiqing) [reply canal/config.go]: I think on or off might be misunderstanding. mysqldump of Mariadb is not compatible with that of MySQL. 
If off is set, it look like --set-gtid-purged=off might be set, but Mariadb is not supported.
So i would prefer to use auto of MySQL, none for Mariadb.
What do you think ?

Comment (siddontang) [reply canal/config.go]: > set_gtid_purged: 'on', 'auto' or 'none' can be set

In the above error message, seem 'on' can be set, but here we only need to set 'none' or 'auto', no 'on'?

Comment (jianhaiqing) [reply canal/config.go]: yes, https://github.com/siddontang/go-mysql/pull/444#discussion_r341884561
canal/canal.go should be fixed. Could you fix it ? Or I fix it by myself. 
Sorry about these trouble

File: canal/canal.go [hunk 7]
```diff
@@ -120,6 +120,23 @@ func NewCanal(cfg *Config) (*Canal, error) {
 	return c, nil
 }
 
+func (c *Canal) validateSetGtidPurged() error {
+	gtidPuged := strings.ToLower(c.cfg.Dump.GtidPurged)
+	if gtidPuged == "none" {
+		return nil
+	} else if gtidPuged == "auto" {
+		res, err := c.Execute(`SHOW GLOBAL VARIABLES LIKE "gtid_mode";`)
```
Comment (WangXiangUSTC) [inline canal/canal.go]: how about adding a function named `GetGtidMode`?

File: canal/canal.go [hunk 8]
```diff
@@ -120,6 +120,23 @@ func NewCanal(cfg *Config) (*Canal, error) {
...
+		if err != nil {
+			return errors.Trace(err)
+		} else if f, _ := res.GetString(0, 1); f != "ON" {
```
Comment (WangXiangUSTC) [inline canal/canal.go]: how about using `strings.ToLower(f) != "on"`?

File: canal/canal_test.go [hunk 9]
```diff
@@ -74,6 +82,126 @@ func (s *canalTestSuite) SetUpSuite(c *C) {
...
+		expectedGtidset, err := mysql.ParseGTIDSet("mysql", tt.expected)
+		if err != nil {
+			log.Fatalf("Gtid:%s failed parsed, err: %v", tt.expected, err)
+		}
+		if !expectedGtidset.Equal(h.gset) {
+			log.Fatalf("expected:%v , but get: %v", expectedGtidset, h.gset)
+		}
+
+		//	c.Assert(expectedGtidset.Equal(h.gset), IsTrue)
+		//	c.Logf("parsed gtidset: %v", h.gset)
```
Comment (WangXiangUSTC) [inline canal/canal_test.go]: can remove these useless code

File: canal/config.go [hunk 10]
```diff
@@ -113,6 +116,8 @@ func NewDefaultConfig() *Config {
 	c.Dump.ExecutionPath = "mysqldump"
 	c.Dump.DiscardErr = true
 	c.Dump.SkipMasterData = false
+	// add default value to disable mysqldump --set-gtid-purged
+	c.Dump.GtidPurged = "none"
```
Comment (WangXiangUSTC) [inline canal/config.go]: how about defining two variable for "none" and "auto"

File: dump/parser.go [hunk 11]
```diff
@@ -54,6 +65,43 @@ func Parse(r io.Reader, h ParseHandler, parseBinlogPos bool) error {
 			return c == '\r' || c == '\n'
 		})
 
+		// parsed gtid set from mysqldump,refer to canal_test.go TestDumperHandler function
```
Comment (WangXiangUSTC) [inline dump/parser.go]:
```suggestion
		// parsed gtid set from mysqldump, refer to canal_test.go TestDumperHandler function
```

File: canal/canal.go [hunk 12]
```diff
@@ -120,6 +120,23 @@ func NewCanal(cfg *Config) (*Canal, error) {
...
+			return errors.Errorf("set-gtid-purged: %s, gtid_mode should be ON, but now is %s", c.cfg.Dump.GtidPurged, f)
+		}
+		return nil
+	}
+
+	return errors.Errorf("set-gtid-purged: none or auto  can be set, current is %s", gtidPuged)
```
Comment (WangXiangUSTC) [inline canal/canal.go]:
```suggestion
	return errors.Errorf("set-gtid-purged: none or auto can be set, current is %s", gtidPuged)
```

File: dump/dump.go [hunk 13]
```diff
@@ -143,6 +149,10 @@ func (d *Dumper) Dump(w io.Writer) error {
 		args = append(args, "--master-data")
 	}
 
+	if d.gtidPurged == "auto" {
+		args = append(args, fmt.Sprintf("--set-gtid-purged=%s", d.gtidPurged))
```
Comment (WangXiangUSTC) [inline dump/dump.go]: don't support MariaDB now?

Comment (jianhaiqing) [reply dump/dump.go]: Yes, this PR add GTID replication for MySQL after myqldump is done, MariaDB can still use file-pos replication.

File: canal/canal.go (hunk 12 above)
Comment (jianhaiqing) [reply canal/canal.go]: OK

File: dump/parser.go (hunk 11 above)
Comment (jianhaiqing) [reply dump/parser.go]: OK

File: canal/config.go (hunk 10 above)
Comment (jianhaiqing) [reply canal/config.go]: GtidPurged is not used that much, I think it's ok right now.

File: canal/canal_test.go (hunk 9 above)
Comment (jianhaiqing) [reply canal/canal_test.go]: OK

File: canal/canal.go (hunk 8 above)
Comment (jianhaiqing) [reply canal/canal.go]: OK

File: canal/canal.go (hunk 7 above)
Comment (jianhaiqing) [reply canal/canal.go]: Nice tips. OK

File: dump/parser.go [hunk 14]
```diff
@@ -54,6 +65,43 @@ func Parse(r io.Reader, h ParseHandler, parseBinlogPos bool) error {
...
+				if m := endUuidSet.FindAllStringSubmatch(line, -1); len(m) == 1 {
+					gset := m[0][1]
+					if err := h.UpdateGtidFromPurged(gset); err != nil {
+						return errors.Trace(err)
+					}
+					gtidDoneParsed = true
+				}
+
+			}
+		}
```
Comment (WangXiangUSTC) [inline dump/parser.go]: these codes want to get all gtid set in dump file, and then update dumpParseHandler's `gset`?
can we make the code more simple?

File: canal/canal.go [hunk 15]
```diff
@@ -120,6 +120,32 @@ func NewCanal(cfg *Config) (*Canal, error) {
...
+		return nil
+	}
+
+	return errors.Errorf("set-gtid-purged: none or auto can be set, current is %s", gtidPuged)
+}
+
+// if MySQL gtid_mode is on, return true, otherwise false
+func (c *Canal) GetGtidMode() (bool, error) {
+	res, err := c.Execute(`SHOW GLOBAL VARIABLES LIKE "gtid_mode";`)
+	if err != nil {
```
Comment (totalo) [inline canal/canal.go]:
```suggestion
	if err == nil {
```

**Comment by GregoryIan** [comment]
@WangXiangUSTC PTAL

File: canal/config.go [hunk 16]
```diff
@@ -34,6 +34,9 @@ type DumpConfig struct {
 	// 'FLUSH TABLES WITH READ LOCK'
 	SkipMasterData bool `toml:"skip_master_data"`
 
+	// set --set-gtid-purged none, auto; none for gtid is disabled or "version too low", auto for gtid_mode=on;
+	GtidPurged string `toml:"set_gtid_purged"`
```
[the remark this answers is not in the document]
Comment (GregoryIan) [reply canal/config.go]: What are the consequences if `--set-gtid-purged` is used on MariaDB?

File: canal/config.go (hunk 10 above)
Comment (GregoryIan) [reply canal/config.go]: Previously the default value was always `auto`,  will it affect other user behaviors after being changed to `none`？

File: dump/parser.go [hunk 17]
```diff
@@ -19,28 +19,73 @@ var (
...
 
 func init() {
+	//SET @@GLOBAL.GTID_PURGED='1638041a-0457-11e9-bb9f-00505690b730:1-429405150';
+	oneGtidExp = regexp.MustCompile("SET @@GLOBAL.GTID_PURGED='(.+)'")
+
+	//SET @@GLOBAL.GTID_PURGED='071a84e8-b253-11e8-8472-005056a27e86:1-76,	// start of multi-gtid set
+	//2337be48-0456-11e9-bd1c-00505690543b:1-7,								// middle gtid set
+	//41d816cd-0455-11e9-be42-005056901a22:1-2';   							// end gtid set
+	mutilGtidStartExp = regexp.MustCompile("SET @@GLOBAL.GTID_PURGED='(.+),")
+	midUuidSet = regexp.MustCompile("(^\\w{8}(-\\w{4}){3}-\\w{12}:\\d+-\\d+),")
```
Comment (GregoryIan) [inline dump/parser.go]: what's meaning of `start`, `middle`, `end` gtid set?

**Comment by GregoryIan** [comment]
I'm not quite understand the purpose of this pr, do  you want to automatically replicate binlog data starting from gtid if gtid is turned on for both the source database and the target data?

File: dump/parser.go (hunk 17 above)
Comment (jianhaiqing) [reply dump/parser.go]: `SET @@GLOBAL.GTID_PURGED='071a84e8-b253-11e8-8472-005056a27e86:1-76,
2337be48-0456-11e9-bd1c-00505690543b:1-7,
41d816cd-0455-11e9-be42-005056901a22:1-2';`
GTID_PURGED might contains multiple gtid sets, so I use start, middle, end, these three regex to match these conditions.

**Comment by jianhaiqing** [comment]
> I'm not quite understand the purpose of this pr, do you want to automatically replicate binlog data starting from gtid if gtid is turned on for both the source database and the target data?

Yes, if replicate start with GTID, it would be easier to continue to replicate after master-slave  switch occurs.

**Comment by jianhaiqing** [comment]
As you know, there might be SLB in front of master or slave, so gtid replication would be better.

**Comment by jianhaiqing** [comment]
Purpose of This PR:
1. if MySQL is GTID_MODE=ON, canal would be start to replicate with GTID after mysqldump is done.
2. Add one more func GtidSet() for parseHandler, to save gtid set after parsing gtid from mysqldump.

**Comment by WangXiangUSTC** [comment]
> Purpose of This PR:
> 
> 1. if MySQL is GTID_MODE=ON, canal would be start to replicate with GTID after mysqldump is done.
> 2. Add one more func GtidSet() for parseHandler, to save gtid set after parsing gtid from mysqldump.

can update thease description in the first comment

File: dump/dump_test.go [hunk 18]
```diff
@@ -119,10 +127,99 @@ func (h *testParseHandler) BinLog(name string, pos uint64) error {
...
+		parseBinlogPos := true
+		for {
+			bytes, _, err := newReader.ReadLine()
+			line := string(bytes)
+			if err == io.EOF {
+				break
+			}
+
+			if parseBinlogPos && !binlogParsed {
+				if m := gtidExp.FindAllStringSubmatch(line, -1); len(m) == 1 {
```
Comment (WangXiangUSTC) [inline dump/dump_test.go]: this unit test just test the regexp `gtidExp`, how about add test for `Parse` function

File: canal/config.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (jianhaiqing) [reply canal/config.go]: remove

File: dump/dump_test.go (hunk 18 above)
Comment (jianhaiqing) [reply dump/dump_test.go]: The regexp gtidExp is used to parse GTID, and GtidSet of dumpParseHandler is used to saveAndUpdate GTID sets which will be used to establish the replication.

Comment (jianhaiqing) [reply dump/dump_test.go]: Add test for Parse already.

File: dump/dump_test.go [hunk 19]
```diff
@@ -9,6 +9,9 @@ import (
 	"strings"
 	"testing"
 
+	"github.com/siddontang/go-log/log"
+	"github.com/siddontang/go-mysql/mysql"
+
```
Comment (WangXiangUSTC) [inline dump/dump_test.go]: please remove this line

File: mysql/mysql_gtid.go [hunk 20]
```diff
@@ -9,8 +9,9 @@ import (
 	"strconv"
 	"strings"
 
+	uuid "github.com/satori/go.uuid"
+
```
Comment (WangXiangUSTC) [inline mysql/mysql_gtid.go]: Seems  don’t need change this

File: canal/dump.go [hunk 21]
```diff
@@ -27,6 +27,15 @@ func (h *dumpParseHandler) BinLog(name string, pos uint64) error {
 	return nil
 }
 
+func (h *dumpParseHandler) GtidSet(gtidsets string) (err error) {
```
Comment (WangXiangUSTC) [inline canal/dump.go]: How about change name to SetGtidSet

File: canal/dump.go [hunk 22]
```diff
@@ -27,6 +27,15 @@ func (h *dumpParseHandler) BinLog(name string, pos uint64) error {
...
+	if h.gset != nil {
+		err = h.gset.Update(gtidsets)
+	} else {
+		h.gset, err = mysql.ParseGTIDSet("mysql", gtidsets)
```
Comment (WangXiangUSTC) [inline canal/dump.go]: Can flavor be mariadb?

File: dump/parser.go [hunk 23]
```diff
@@ -19,18 +19,21 @@ var (
...
 var useExp *regexp.Regexp
 var valuesExp *regexp.Regexp
+var gtidExp *regexp.Regexp
 
 func init() {
 	binlogExp = regexp.MustCompile("^CHANGE MASTER TO MASTER_LOG_FILE='(.+)', MASTER_LOG_POS=(\\d+);")
 	useExp = regexp.MustCompile("^USE `(.+)`;")
 	valuesExp = regexp.MustCompile("^INSERT INTO `(.+?)` VALUES \\((.+)\\);$")
+	//SET @@GLOBAL.GTID_PURGED='1638041a-0457-11e9-bb9f-00505690b730:1-429405150';
+	gtidExp = regexp.MustCompile("(\\w{8}(-\\w{4}){3}-\\w{12}:\\d+-\\d+)")
```
Comment (WangXiangUSTC) [inline dump/parser.go]: How about add a comment this pattern only match MySQL grid?

File: dump/dump_test.go (hunk 19 above)
Comment (jianhaiqing) [reply dump/dump_test.go]: these two packages are used in unit test

File: canal/dump.go (hunk 22 above)
Comment (jianhaiqing) [reply canal/dump.go]: it would be better to add one more parameter for Dumper, but at the moment, it's ok

File: dump/parser.go (hunk 23 above)
Comment (jianhaiqing) [reply dump/parser.go]: if someone use MariaDB and it should know `SET @@GLOBAL.GTID_PURGED=` not for MariaDB, only `SET GLOBAL gtid_slave_pos` is.

File: canal/dump.go (hunk 21 above)
Comment (jianhaiqing) [reply canal/dump.go]: its naming style follows Binlog func, I think it's ok

File: mysql/mysql_gtid.go (hunk 20 above)
Comment (jianhaiqing) [reply mysql/mysql_gtid.go]: it's added by IDE, i haven't figure out why. If i remove it directly, it will fail to compile.

**Comment by WangXiangUSTC** [comment]
LGTM

File: canal/sync.go [hunk 24]
```diff
@@ -16,7 +16,7 @@ import (
 
 func (c *Canal) startSyncer() (*replication.BinlogStreamer, error) {
 	gset := c.master.GTIDSet()
-	if gset == nil {
+	if gset == nil || (gset != nil && gset.String() == "") {
```
Comment (siddontang) [inline canal/sync.go]:
```
(gset != nil && gset.String() == "")
```
seem here we can use `gset.String()` directly beause the first condition has already checked gset is nil.

Comment (jianhaiqing) [reply canal/sync.go]: Yes

**Pull request merged**